### PR TITLE
[HB-6524] Public NuGet Distribution

### PR DESCRIPTION
**Pull request by kushG**
# Title
[HB-6524] Public Nuget Distribution

# Description

Enables NuGet distribution of Chartboost Mediation Unity SDK 

To accommodate Nuget distribution and UPM distribution together within this same repository the project structure is restructured as below
```
com.chartboost.mediation                 -- all the sdk code
com.chartboost.mediation.demo            -- all the code for demo
com.chartboost.mediation.demo.upm        -- unity project that integrates demo locally and sdk through NPM/UPM
com.chartboost.mediation.demo.nuget      -- unity project that integrates demo locally and sdk through Nuget/NugetForUnity
```

CI tasks run on both `com.chartboost.mediation.demo.upm` and `com.chartboost.mediation.demo.nuget` with the latter requiring a nuget packaging [job](https://github.com/ChartBoost/chartboost-mediation-unity-sdk/blob/1bceba8acce4dc7dca6f9fd7e401cd1f677d8a77/.github/workflows/status.yml#L74) as a prerequisite. 

## Type of change

- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)
- [ ] Refactor / Maintenance (refactoring code to be cleaner/easier to maintain)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

[HB-6524]: https://chartboost.atlassian.net/browse/HB-6524?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/actions/compile-unity/action.yml [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+runs:
+  using: "composite"
+  
+  steps:
+    - name: Setup Unity Editor
+      id: unity-install
+      uses: ./.github/actions/chartboost-mediation-unity-ci/setup-unity
+      with:
+        unity-version: ${{ env.UNITY_VERSION }}
+        unity-version-changeset: 8c66806a0c04
```
Comment (SCastanedaMunoz) [inline .github/actions/compile-unity/action.yml:30]: You should be able to specify this value.

File: com.chartboost.mediation.demo.nuget/Assets/Plugins/Android/AndroidManifest.xml [hunk 2]
```diff
@@ -14,5 +14,7 @@
       <meta-data android:name="unityplayer.UnityActivity" android:value="true" />
       <meta-data android:name="unityplayer.ForwardNativeEventsToDalvik" android:value="true" />
     </activity>
+    <meta-data android:name="com.google.android.gms.ads.APPLICATION_ID" android:value="ca-app-pub-6548817822928201~9780573322" />
+    <meta-data android:name="applovin.sdk.key" android:value="wy1dIpr2F2-9fqTT2ihBprfAvPyfd6qm_81WkbsZC6Fl31vaW4aiDFxR52R89x98WWNfPmTIGMInIvffFTOxLz" />
```
Comment (SCastanedaMunoz) [inline com.chartboost.mediation.demo.nuget/Assets/Plugins/Android/AndroidManifest.xml:18]: This should not be commited.

File: com.chartboost.mediation.demo.nuget/Assets/com.chartboost.mediation/Editor/selections.json [hunk 3]
```diff
@@ -0,0 +1,100 @@
+{
+  "mediationVersion": "4.6.0",
```
Comment (SCastanedaMunoz) [inline com.chartboost.mediation.demo.nuget/Assets/com.chartboost.mediation/Editor/selections.json:2]: Shouldn't this be using 4.7.0?

File: com.chartboost.mediation.demo.upm/Assets/Plugins/Android/AndroidManifest.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    <activity android:name="com.unity3d.player.UnityPlayerActivity" android:theme="@style/UnityThemeSelector">
+      <intent-filter>
+        <action android:name="android.intent.action.MAIN" />
+        <category android:name="android.intent.category.LAUNCHER" />
+      </intent-filter>
+      <meta-data android:name="unityplayer.UnityActivity" android:value="true" />
+      <meta-data android:name="unityplayer.ForwardNativeEventsToDalvik" android:value="true" />
+    </activity>
+    <meta-data android:name="com.google.android.gms.ads.APPLICATION_ID" android:value="ca-app-pub-6548817822928201~9780573322" />
+    <meta-data android:name="applovin.sdk.key" android:value="wy1dIpr2F2-9fqTT2ihBprfAvPyfd6qm_81WkbsZC6Fl31vaW4aiDFxR52R89x98WWNfPmTIGMInIvffFTOxLz" />
```
Comment (SCastanedaMunoz) [inline com.chartboost.mediation.demo.upm/Assets/Plugins/Android/AndroidManifest.xml:18]: This 2 fields should not be commited, the AndroidManifest in the demo apps should contain a clean version of it so people can use whatever keys/setup they need.

**Review by SCastanedaMunoz — COMMENTED** [review]
Look at comments.

File: com.chartboost.mediation.demo.upm/Assets/Plugins/Android/AndroidManifest.xml (hunk 4 above)
Comment (kushG) [reply com.chartboost.mediation.demo.upm/Assets/Plugins/Android/AndroidManifest.xml:18]: oh yeah this was by mistake

File: .github/actions/compile-unity/action.yml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+      id: notify-slack
+      uses: ./.github/actions/chartboost-mediation-unity-ci/notify-slack
+      with:
+        slack-webhook-url: ${{ inputs.slack-webhook }}
+        workflow-run-id: ${{ github.run_id }}
+        workflow-name: ${{ github.workflow }}
+        workflow-url: "https://github.com/${{ github.repository }}/actions/runs/${{github.run_id}}"
+
+
+
```
Comment (SCastanedaMunoz) [inline .github/actions/compile-unity/action.yml:128]: Can we remove this extra whitespace? just leave a single empty line.

File: .github/actions/compile-unity/action.yml [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  username:
+    description: Unity account username associated with license serial number. Use only if activating or returning license. Using Github actions secrets is recommended.
+    required: true
+  password:
+    description: Unity account password associated with license serial number. Use only if activating or returning license. Using Github actions secrets is recommended.
+    required: true
+  slack-webhook:
+    description: Webhook url for posting on slack. Using Github actions secrets is recommended.
+    required: true
+
```
Comment (SCastanedaMunoz) [inline .github/actions/compile-unity/action.yml:19]: Remove extra line

File: .github/workflows/status.yml [hunk 7]
```diff
@@ -69,110 +79,98 @@ jobs:
...
+
+    - name: Upload Nuget Package
+      continue-on-error: true      
+      run : |
+        git config user.name ${{ secrets.GIT_USER }}
+        git config user.email ${{ secrets.GIT_EMAIL }}
+        git add Nuget/Chartboost.CSharp.Mediation.Unity*.nupkg   
+        git commit -m "Updating Chartboost.CSharp.Mediation.Unity.nupkg"
+        git push --force
+
```
Comment (SCastanedaMunoz) [inline .github/workflows/status.yml:117]: Remove extra line

File: .github/workflows/status.yml [hunk 8]
```diff
@@ -69,110 +79,98 @@ jobs:
...
+        token: ${{ secrets.GITHUBSERVICETOKEN }}
 
-    - name: Return Unity License
-      if: always()
-      id: return-unity
-      uses: ./.github/actions/chartboost-mediation-unity-ci/run-unity
+    - name: Setup DotNet
+      uses: actions/setup-dotnet@v3
+
+    - name: Install/Restore Chartboost Mediation Unity SDK Nuget Package
```
Comment (SCastanedaMunoz) [inline .github/workflows/status.yml:160]:
```suggestion
    - name: Install/Restore Chartboost Mediation Unity SDK NuGet Package
```

File: .github/workflows/status.yml [hunk 9]
```diff
@@ -69,110 +79,98 @@ jobs:
...
-    - name: Check for Compiler Errors in iOS
-      id: unity-compile-ios
-      uses: ./.github/actions/chartboost-mediation-unity-ci/run-unity
-      with:
-        unity-executable-path: ${{ env.UNITY_EXE_PATH }}
-        project-path: ${{ env.PROJECT_PATH }}
-        target-platform: ios
+  
+  compile-unity-nuget:
+    name: Compile Nuget Unity Project
```
Comment (SCastanedaMunoz) [inline .github/workflows/status.yml:145]:
```suggestion
    name: Compile NuGet Unity Project
```

File: .github/workflows/status.yml [hunk 10]
```diff
@@ -69,110 +79,98 @@ jobs:
...
+        # trigger another workflow and thus avoid recursive workflow trigger
+        # If we use `secrets.GITHUBSERVICETOKEN` instead it will overwrite the above behaviour and will not prevent
+        # recursive workflow
+        # https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow
+        token: ${{ secrets.GITHUB_TOKEN }}
+
+    - name: Generate Nuget Package
+      run : nuget pack ./com.chartboost.mediation/Chartboost.CSharp.Mediation.Unity.nuspec  -OutputDirectory ./Nuget -NoDefaultExcludes
+
+    - name: Upload Nuget Package
```
Comment (SCastanedaMunoz) [inline .github/workflows/status.yml:109]:
```suggestion
    - name: Upload NuGet Package
```

File: .github/workflows/status.yml [hunk 11]
```diff
@@ -69,110 +79,98 @@ jobs:
...
+      with:
+        ref: ${{ steps.extract_branch.outputs.branch }}
+        # Using `secrets.GITHUB_TOKEN` ensures that any commit(s) pushed from within this workflow will not 
+        # trigger another workflow and thus avoid recursive workflow trigger
+        # If we use `secrets.GITHUBSERVICETOKEN` instead it will overwrite the above behaviour and will not prevent
+        # recursive workflow
+        # https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow
+        token: ${{ secrets.GITHUB_TOKEN }}
+
+    - name: Generate Nuget Package
```
Comment (SCastanedaMunoz) [inline .github/workflows/status.yml:106]:
```suggestion
    - name: Generate NuGet Package
```

File: .github/actions/compile-unity/action.yml (hunk 6 above)
Comment (SCastanedaMunoz) [inline .github/actions/compile-unity/action.yml:19]:
```suggestion
```

File: .github/actions/compile-unity/action.yml [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+      if: ${{ failure() }}
+      id: notify-slack
+      uses: ./.github/actions/chartboost-mediation-unity-ci/notify-slack
+      with:
+        slack-webhook-url: ${{ inputs.slack-webhook }}
+        workflow-run-id: ${{ github.run_id }}
+        workflow-name: ${{ github.workflow }}
+        workflow-url: "https://github.com/${{ github.repository }}/actions/runs/${{github.run_id}}"
+
+
```
Comment (SCastanedaMunoz) [inline .github/actions/compile-unity/action.yml:127]:
```suggestion
```

File: .github/workflows/status.yml [hunk 13]
```diff
@@ -25,25 +25,35 @@ on:
...
 
 jobs:
-  compile:
+  android-bridge:
     name: Compile Android Bridge
     runs-on: ubuntu-latest
 
-    steps:
+    steps:   
+    - name: Extract branch name
```
Comment (SCastanedaMunoz) [inline .github/workflows/status.yml:42]:
```suggestion
    - name: Extract Branch Name
```

File: .github/workflows/status.yml [hunk 14]
```diff
@@ -69,110 +79,98 @@ jobs:
...
+        git push --force
+  
+  nuget:
+    name: Build Nuget package
+    runs-on: ubuntu-latest
 
-  tests:
-    name: Unity Actions & Unit Testing
+    steps:
+    - name: Extract branch name
```
Comment (SCastanedaMunoz) [inline .github/workflows/status.yml:90]:
```suggestion
    - name: Extract branch name
```
```suggestion
    - name: Extract Branch Name
```

File: .github/workflows/status.yml [hunk 15]
```diff
@@ -69,110 +79,98 @@ jobs:
...
+
+    - name: Generate Nuget Package
+      run : nuget pack ./com.chartboost.mediation/Chartboost.CSharp.Mediation.Unity.nuspec  -OutputDirectory ./Nuget -NoDefaultExcludes
+
+    - name: Upload Nuget Package
+      continue-on-error: true      
+      run : |
+        git config user.name ${{ secrets.GIT_USER }}
+        git config user.email ${{ secrets.GIT_EMAIL }}
+        git add Nuget/Chartboost.CSharp.Mediation.Unity*.nupkg
```
Comment (SCastanedaMunoz) [inline .github/workflows/status.yml:114]: The folder should be named NuGet as that is the official spelling.

File: .github/workflows/status.yml [hunk 16]
```diff
@@ -69,110 +79,98 @@ jobs:
...
-        unity-version-changeset: 8c66806a0c04
-        unity-modules:
-          android
-          ios
-
-    - name: Activate Unity License
-      id: activate-unity
-      uses: ./.github/actions/chartboost-mediation-unity-ci/run-unity
+        
+    - name: compile-unity
```
Comment (SCastanedaMunoz) [inline .github/workflows/status.yml:132]: Can we rename this step like we talked about?

**Review by SCastanedaMunoz — COMMENTED** [review]
Look at comments.

File: com.chartboost.mediation/Dependencies/ChartboostMediationDependencyPackageInstaller.cs [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+namespace Dependencies
+{
+    /// <summary>
+    /// Helper class that allows installation of any additional dependencies that are available to install using Unity Package Manager(UPM)
+    /// This would allow installation of UPM based packages even when sdk is not installed using UPM
+    /// </summary>
+    [InitializeOnLoad]
+    public static class ChartboostMediationDependencyPackageInstaller
+    {
+        private const string NewtonSoftJsonDependency = "com.unity.nuget.newtonsoft-json@3.2.1";
```
Comment (SCastanedaMunoz) [inline com.chartboost.mediation/Dependencies/ChartboostMediationDependencyPackageInstaller.cs:13]: Can we separate the version number from this into its own variable?

File: com.chartboost.mediation/Runtime/Utilities/ChartboostMediationUtils.cs [hunk 18]
```diff
@@ -1,3 +1,4 @@
+using System.IO;
```
Comment (SCastanedaMunoz) [inline com.chartboost.mediation/Runtime/Utilities/ChartboostMediationUtils.cs:1]: Is this needed?

**Review by SCastanedaMunoz — COMMENTED** [review]
Look at comments

**Review by SCastanedaMunoz — APPROVED** [review]

**Pull request merged**